src/main/java/uk/ac/ebi/subs/ena/processor/ENAProcessor.java

```diff
@@ -169,5 +190,89 @@ private String centerName(SubmissionEnvelope submissionEnvelope) {
         return paramMap;
     }
 
+    private Study setDefaultStudyReleaseDate(SubmissionEnvelope submissionEnvelope, Study study) {
+        Date subDate = submissionEnvelope.getSubmission().getSubmissionDate();
+
+        if (subDate != null && !study.isAccessioned() && study.getReleaseDate() == null) {
+            study.setReleaseDate(convertSubmissionDate(submissionEnvelope.getSubmission().getSubmissionDate()));
+        }
+
+        return study;
+    }
+
+    private Sample setDefaultSampleReleaseDate(SubmissionEnvelope submissionEnvelope, Sample sample) {
+        Date subDate = submissionEnvelope.getSubmission().getSubmissionDate();
 
+        if (subDate != null && !sample.isAccessioned() && sample.getReleaseDate() == null) {
+            sample.setReleaseDate(convertSubmissionDate(submissionEnvelope.getSubmission().getSubmissionDate()));
+        }
+
+        return sample;
+    }
+
+    private LocalDate convertSubmissionDate(Date submissionDate) {
+        return LocalDate.ofInstant(Instant.ofEpochMilli(submissionDate.getTime()), ZoneId.of("UTC"));
+    }
+
+    /**
+     * Organize the action services and their submittables into groups with similar release dates.
+     *
+     * @param serviceSubmittableMap
+     * @return
+     */
+    private Map<Optional<LocalDate>, Map<Class<? extends ActionService>, Object>> createReleaseDateGroups(
+            Map<Class<? extends ActionService>, Object> serviceSubmittableMap) {
+
+        //Map with date as the key and its group as value.
+        Map<Optional<LocalDate>, Map<Class<? extends ActionService>, Object>> groups = new LinkedHashMap<>();
+
+        serviceSubmittableMap.entrySet().stream().forEach(serviceSubmittableEntry -> {
+
+            Class<? extends ActionService> service = serviceSubmittableEntry.getKey();
+            Object serviceSubmittable = serviceSubmittableEntry.getValue();
+
+            groupByReleaseDate(serviceSubmittable).entrySet().stream().forEach(dateSubmittableEntry -> {
+
+                Optional<LocalDate> date = dateSubmittableEntry.getKey();
+                Object dateSubmittable = dateSubmittableEntry.getValue();
+
+                Map<Class<? extends ActionService>, Object> group = groups.get(date);
```

Comment:
I am not sure why don't you simple use `groups.put(date, Map.of(service, dateSubmittable))`
From API doc `LinkedhashMap` - `put`method: Associates the specified value with the specified key in this map. If the map previously contained a mapping for the key, the old value is replaced.

```suggestion
                groups.put(date, Map.of(service, dateSubmittable));
```

Reply:
if i got this right then what you're suggesting will replace an existing group from the groups map. which is not what we want. Instead, if a group is present already then we take it and update it.

please let me know if i got it wrong?

Reply:
I think with this line you change the reference anyway: `group.put(service, dateSubmittable);`. Am I wrong?

Reply:
that line changes updates the group. i think your suggestion was about line 239. that would have always replaced an existing group.

Reply:
Ok, I got it

But what about this:

```suggestion
                Map<Class<? extends ActionService>, Object> group = groups.get(date);
                groups.put(date, group == null ? Map.of(service, dateSubmittable) : group.put(service, dateSubmittable));
```

Reply:
Ok, got it. I was wrong